server/src/test/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshotsTest.java

```diff
@@ -0,0 +1,106 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.index.snapshots.blobstore;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.elasticsearch.repositories.blobstore.BlobStoreRepository.INDEX_SHARD_SNAPSHOTS_FORMAT;
+
+import java.lang.management.ManagementFactory;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.lucene.util.Version;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+import org.elasticsearch.index.store.StoreFileMetadata;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Test;
+
+import com.sun.management.ThreadMXBean;
+
+public class BlobStoreIndexShardSnapshotsTest extends ESTestCase {
+
+
+    @Test
+    public void test_bwc_streaming() throws Exception {
+        BlobStoreIndexShardSnapshots blobStoreIndexShardSnapshots = prepareData();
+        BytesStreamOutput out = new BytesStreamOutput();
+        blobStoreIndexShardSnapshots.writeTo(out);
+
+        StreamInput in = out.bytes().streamInput();
+        BlobStoreIndexShardSnapshots fromStream = new BlobStoreIndexShardSnapshots(in);
+
+        //TODO: implement equals or compare per-field, also, try different versions
+        assertThat(blobStoreIndexShardSnapshots).isEqualTo(fromStream);
+    }
+
+    @Test
+    public void test_read_from_stream_allocations() throws Exception {
+        BytesReference bytesReference = INDEX_SHARD_SNAPSHOTS_FORMAT.serialize(prepareData(), "dummyBlobName", true);
+
+        ThreadMXBean threadMXBean = ManagementFactory.getPlatformMXBean(ThreadMXBean.class);
+        long threadId = Thread.currentThread().threadId();
+        long allocatedBytesBegin = threadMXBean.getThreadAllocatedBytes(threadId);
+
+        INDEX_SHARD_SNAPSHOTS_FORMAT.deserialize("dummyBlobname", writableRegistry(), xContentRegistry(), bytesReference);
+
+        long allocatedBytesAfter = threadMXBean.getThreadAllocatedBytes(threadId);
+        long allocatedBytes = allocatedBytesAfter - allocatedBytesBegin;
+
+        // Running this test on a state before https://github.com/crate/crate/commit/8f941f829d1fe4193f968af9a96bd3e120708fd4
+        // (when BlobStoreIndexShardSnapshots used to implement toXContent)
+        // gets 1_195_304, using this as a reference upper bound.
+        assertThat(allocatedBytes).isLessThanOrEqualTo(1_195_304);
+
+        // 5.10.1 with regression and even without compression fix returns 748208L.
+        // with compression fix: fluctuates around 574136L
+
+        // Regressed version can't be better, wrong approach to test??
+    }
```

Comment:
This is confusing. 

If XContent had less allocations why only 5.10 have OOM?

Also, judging by recordings, 5.10.2 and especially 5.10.1 do more stuff. 

Either test is invalid or fix is wrong?

Reply:
@mfussenegger could you please take a look at recordings and check if an update i this PR makes sense?

To me it looks like overall 5.10.1 -> 5.10.2 -> this PR are getting better but not sure that OOM would be addressed.

Maybe I should compare this PR with 5.10.1 and 5.10.2?

Reply:
UPD: Test was not covering a problematic scenario:` SnapshotFiles.indexFiles` overlapping across different instnces.

5.9 used to write/read in a way that FileInfo instances were re-used.
In 5.10 we streamed all structures separately and such had different instances even for "logical overlap".